share-coordinator/src/main/java/org/apache/kafka/coordinator/share/PersisterStateBatchCombiner.java

```diff
@@ -0,0 +1,394 @@
+        }
+    }
+
+    /**
+     * Algorithm: Merge current state batches and new batches into a single non-overlapping batch list.
+     * Input: batchesSoFar, newBatches, startOffset
+     * Output: combined list with non-overlapping batches (finalBatchList)
+     * <p>
+     * - Add both currentBatches and newBatches into a single list combinedBatchList
+     * - Remove/prune any batches from the combinedBatchList:
+     * -    if batch.lastOffset < startOffset then remove batch from combinedBatchList
+     * -    else if batch.firstOffset > startOffset then we will keep the batch
+     * -    else if batch.firstOffset <= startOffset <= batch.lastOffset then keep [startOffset, batch.lastOffset] part only and discard rest.
+     * - create a treeset sortedBatches using pruned combinedBatchList
+     * - do repeat until all batches in sortedSet are non-overlapping:
+     * -    find first 2 overlapping batches in sortedBatches set, say, prev and candidate.
+     * -    remove any non-overlapping batches from sortedBatches encountered during the find operation and add them to a finalBatchList
+     * -    based on various conditions of offset overlap and batch state differences combine the batches or
+     *      create new batches, if required, and add to the sortedSet.
+     * - done
+     * - return the finalBatchList
+     *
+     * @return list of {@link PersisterStateBatch} representing non-overlapping combined batches
+     */
+    public List<PersisterStateBatch> combineStateBatches() {
+        pruneBatches();
+        mergeBatches();
+        return finalBatchList;
+    }
+
+    private void mergeBatches() {
+        if (combinedBatchList.size() < 2) {
+            finalBatchList = combinedBatchList;
+            return;
+        }
+
+        sortedBatches = new TreeSet<>(combinedBatchList);
+
+        BatchOverlapState overlapState = getOverlappingState();
+
+        while (overlapState != BatchOverlapState.EMPTY) {
+            PersisterStateBatch prev = overlapState.prev();
+            PersisterStateBatch candidate = overlapState.candidate();
+
+            // remove both previous and candidate for easier
+            // assessment about adding batches to sortedBatches
+            sortedBatches.remove(prev);
+            sortedBatches.remove(candidate);
+
+            if (compareBatchState(candidate, prev) == 0) {  // same state and overlap or contiguous
+                // overlap and same state (prev.firstOffset <= candidate.firstOffset) due to sort
+                // covers:
+                // case:        1        2          3            4          5           6          7 (contiguous)
+                // prev:        ------   -------    -------      -------   -------   --------    -------
+                // candidate:   ------   ----       ----------     ---        ----       -------        -------
+                handleSameStateOverlap(prev, candidate);
+            } else { // diff state and non-contiguous overlap
```

Comment:
Hmm, if we reach here, it's possible for prev and candidate to be contiguous, right?

Reply:
No it's not. 
getOverlappingState returns continuous pair ONLY IF the pair same delivery count and state.

The condition in the if block will capture continuous case for same state.

If it reaches else it has to be overlapping and different state. The getOverlappingState does not return continuous and different state pairs.

In fact, it updates the treeset by removing any which are found.